appengine/src/main/java/com/google/googleinterns/gscribe/services/impl/SpreadsheetServiceImpl.java

```diff
@@ -0,0 +1,271 @@
+     * Calls parseSpreadsheet function to parse the spreadsheet
+     * If first time gets 401 error then refresh the token as access token might have expired
+     *
+     * @param user          ( user object containing tokens )
+     * @param spreadsheetID ( spreadsheet ID of spreadsheet to be read )
+     * @param range         ( range to be read from spreadsheet )
+     * @return ValueRange object
+     * @throws GeneralSecurityException,IOException ( thrown by NetHttpTransport, GoogleClientSecrets, GoogleTokenResponse or by invalid credentials file  )
+     * @throws InvalidDatabaseDataException         ( when the token received from database is invalid or inconsistent with user )
+     * @throws InvalidRequestException              ( when unable to parse the spreadsheet )
+     */
+    @Override
+    public ValueRange parseSpreadsheetRequest(User user, String spreadsheetID, String range) throws GeneralSecurityException, InvalidDatabaseDataException, InvalidRequestException, IOException {
+        try {
+            return parseSpreadsheet(user, spreadsheetID, range);
+        } catch (GoogleJsonResponseException e) {
+            if (e.getStatusCode() == 401) {
+                tokenService.refreshToken(user);
+                userTokenDao.insertUserToken(user);
+                try {
+                    return parseSpreadsheet(user, spreadsheetID, range);
+                } catch (GoogleJsonResponseException ex) {
+                    throw new InvalidRequestException("Unable to parse Spreadsheet");
+                }
+            } else {
+                throw new InvalidRequestException("Unable to parse Spreadsheet");
+            }
+        }
+    }
+
+    /**
+     * Called to submit response to the responses spreadsheet
+     * Forms the response row to be added to spreadsheet having ( start time, student roll number, answers ..., '-' for final points )
+     * Append the responses to the spreadsheet
+     *
+     * @param examInstance ( exam instance object for the exam taken )
+     * @param user         ( paper setter user object to access spreadsheet )
+     * @param examMetadata ( exam metadata for exam to get spreadsheet id )
+     * @throws GeneralSecurityException,IOException ( thrown by NetHttpTransport, GoogleClientSecrets, GoogleTokenResponse or by invalid credentials file  )
+     */
+    private void addResponseToSheet(ExamInstance examInstance, User user, ExamMetadata examMetadata) throws GeneralSecurityException, IOException {
+        final NetHttpTransport HTTP_TRANSPORT = GoogleNetHttpTransport.newTrustedTransport();
+        final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+        Credential credential = new Credential(BearerToken.authorizationHeaderAccessMethod()).setAccessToken(user.getAccessToken());
+        Sheets service = new Sheets.Builder(HTTP_TRANSPORT, JSON_FACTORY, credential).setApplicationName("Gscribe").build();
+
+        List<Object> response = new ArrayList<>();
```

Comment:
Is it absolutely required to have List<Object>? Can't we have List<String>?

Reply:
Basically in the ValueRange setValues function, it takes as argument the List<List<Object>> object. So due to the fixation of List<Object> I had to use Object only and not the string. Otherwise it gives error.